engine/src/main/java/com/datatorrent/stram/client/StramClientUtils.java

```diff
@@ -519,28 +520,77 @@ public static FileSystem newFileSystemInstance(Configuration conf) throws IOExce
     }
   }
 
+  /**
+   * Helper function used by both getApexDFSRootDir and getDTDFSRootDir to process dfsRootDir as per the required logic
+   *
+   * @param fs
+   * @param conf
+   * @param dfsRootDir
+   * @param useImpersonated
+   * @param processRelativePath
+   * @return
+   */
+  private static Path processDFSRootDirHelper(FileSystem fs, Configuration conf, String dfsRootDir, boolean useImpersonated,
+      boolean processRelativePath)
+  {
+    try {
+      String userShortName = useImpersonated ?
+          UserGroupInformation.getCurrentUser().getShortUserName() : UserGroupInformation.getLoginUser().getShortUserName();
+      boolean userNameReplaced = false;
+      if (dfsRootDir.contains(DT_DFS_USER_NAME)) {
+        dfsRootDir = dfsRootDir.replace(DT_DFS_USER_NAME, userShortName);
+        conf.set(DT_DFS_ROOT_DIR, dfsRootDir);
+        userNameReplaced = true;
+      }
+      URI uri = new URI(dfsRootDir);
+      if (uri.isAbsolute()) {
+        return new Path(uri);
+      }
+      if (processRelativePath && !userNameReplaced && dfsRootDir.startsWith("/") == false) {
+        dfsRootDir = "/user/" + userShortName + "/" + dfsRootDir;
+      }
+    } catch (IOException ex) {
+      LOG.warn("Error getting user login name {}", dfsRootDir, ex);
+    } catch (URISyntaxException ex) {
+      LOG.warn("{} is not a valid URI. Using the default filesystem to construct the path", dfsRootDir, ex);
+    }
+    return new Path(fs.getUri().getScheme(), fs.getUri().getAuthority(), dfsRootDir);
+  }
+
+  /**
+   * This gets the DFS Root dir to be used at runtime by Apex applications as per the following logic:
+   * Value of apex.app.dfsRootDirectory is referred to as Apex-root-dir below.
+   * A "user" refers to either impersonating or impersonated user:
+   *    If apex.application.path.impersonated is true then use impersonated user else impersonating user.
+   *
+   * <ul>
+   * <li> if Apex-root-dir is blank then just call getDTDFSRootDir to get the old behavior
+   * <li> if Apex-root-dir value has %USER_NAME% in the string then replace it with the user's name, else use the absolute path as is.
+   * <li> if Apex-root-dir value is a relative path then append it to the user's home directory.
+   * </ul>
+   *
+   * @param fs FileSystem object for HDFS file system
+   * @param conf  Configuration object
+   * @return
+   * @throws IOException
+   */
+  public static Path getApexDFSRootDir(FileSystem fs, Configuration conf) throws IOException
+  {
+    String apexDfsRootDir = conf.get(APEX_APP_DFS_ROOT_DIR);
+    if (StringUtils.isBlank(apexDfsRootDir)) {
+      return getDTDFSRootDir(fs, conf);
+    }
+    boolean useImpersonated = conf.getBoolean(StramUserLogin.DT_APP_PATH_IMPERSONATED, false);
+    return processDFSRootDirHelper(fs, conf, apexDfsRootDir, useImpersonated, true);
```

Comment:
Actually I did think of that but there is a minor twist here. The method UserGroupInformation.getXXXUser().getShortUserName() throws IOException and the original code (of getDTDFSRootDir) caught it and did default processing with whatever value of dfsRootDir. That logic will have to change or introduce some kind of code duplication.